File: Algorithms/Dijkstra.py
# really hate python for this bad implement

from math import inf

class Vertex:
    def __init__(self, id):
        self.id = id
        self.neighbors = []

    def add_neighbor(self, id, weight):
        self.neighbors.append({
            'id': id,
            'weight': weight
        })

    def get_weight(self, id):
        for neighbor in neighbors:
            if neighbor['id'] == id:
                return neighbor['weight']

        return None


class Graph:
    def __init__(self):
        self.vertices = []
        self.lastId = 0

    def add_vertex(self):
        vertex = Vertex(self.lastId)
        self.vertices.append(vertex)
        self.lastId = self.lastId + 1

        return vertex

    def get_vertex(self, id):
        for vertex in self.vertices:
            if str(vertex.id) == str(id):
                return vertex

        return None

    def add_connection(self, source, target, weight):
        vertex = self.get_vertex(source)
        if vertex is not None:
            vertex.add_neighbor(target, weight)


g = Graph()
def createVertices():
    a = g.add_vertex()
    b = g.add_vertex()
    c = g.add_vertex()
    d = g.add_vertex()
    e = g.add_vertex()
    f = g.add_vertex()

    g.add_connection(a.id, b.id, 7)
    g.add_connection(a.id, c.id, 9)
    g.add_connection(a.id, f.id, 14)
    g.add_connection(b.id, c.id, 10)
    g.add_connection(b.id, d.id, 15)
    g.add_connection(b.id, a.id, 7)
    g.add_connection(c.id, d.id, 11)
    g.add_connection(c.id, f.id, 2)
    g.add_connection(c.id, a.id, 9)
    g.add_connection(c.id, b.id, 10)
    g.add_connection(d.id, b.id, 15)
    g.add_connection(d.id, c.id, 11)
    g.add_connection(e.id, f.id, 9)
    g.add_connection(e.id, d.id, 6)
    g.add_connection(f.id, e.id, 9)
    g.add_connection(f.id, c.id, 2)
    g.add_connection(f.id, a.id, 14)

createVertices()

def findItem(items, id):
    for item in items:
        if item['id'] == id:
            return item

    return None

def dijkstra(g, source, target):
    queue = []
    dist = []
    prev = []

    # initializing
    for vertex in g.vertices:
        dist.append({
            'id': vertex.id,
            'dist': inf
        })
        prev.append({
            'id': vertex.id,
            'prev': None
        })
        queue.append(vertex)

    # set source node dist 0
    findItem(dist, source.id)['dist'] = 0

    # algorithm
    while len(queue):
        # finding a vertex that has min dist value
        # and call it u
        u = dist[0]
        uIndex = 0
        for i in range(len(queue)):
            found = findItem(dist, queue[i].id)
            if found['dist'] < u['dist']:
                u = found
                uIndex = i
        u = queue[uIndex]
        del queue[uIndex]

        # get neighbors
        for neighbor in u.neighbors:
            found = findItem(dist, u.id)
            alt = found['dist'] + neighbor['weight']
            if alt < findItem(dist, neighbor['id'])['dist']:
                found['dist'] = alt
                findItem(prev, neighbor['id'])['prev'] = u

dijkstra(g, g.get_vertex('0'), g.get_vertex('4'))

File: Algorithms/dijkstra.js
class Vertex {
  constructor(id) {
    this.id = id;
    this.neighbors = [];
    this.dist = Infinity;
    this.prev = undefined;
  }

  addNeighbor(id, weight) {
    this.neighbors.push({
      id,
      weight,
    });
  }

  getWeight(id) {
    return this.neighbors.find(vertex => vertex.id === id).weight;
  }

  reset() {
    this.dist = Infinity;
    this.prev = undefined;
  }
}

class Graph {
  constructor() {
    this.vertices = [];
    this.lastId = 0;
  }

  addVertex() {
    const vertex = new Vertex(this.lastId);
    this.vertices.push(vertex);
    this.lastId = this.lastId + 1;

    return vertex;
  }

  addConnection(source, target, weight) {
    this.vertices.find(vertex => vertex.id === source).addNeighbor(target, weight);
  }

  getVertex(id) {
    return this.vertices.find(vertex => vertex.id === id);
  }

  reset() {
    for (const vertex of this.vertices) {
      vertex.reset();
    }
  }
}

const graph = new Graph();

const a = graph.addVertex();
const b = graph.addVertex();
const c = graph.addVertex();
const d = graph.addVertex();
const e = graph.addVertex();
const f = graph.addVertex();

graph.addConnection(a.id, b.id, 7);
graph.addConnection(a.id, c.id, 9);
graph.addConnection(a.id, f.id, 14);
graph.addConnection(b.id, c.id, 10);
graph.addConnection(b.id, d.id, 15);
graph.addConnection(b.id, a.id, 7);
graph.addConnection(c.id, d.id, 11);
graph.addConnection(c.id, f.id, 2);
graph.addConnection(c.id, a.id, 9);
graph.addConnection(c.id, b.id, 10);
graph.addConnection(d.id, b.id, 15);
graph.addConnection(d.id, c.id, 11);
graph.addConnection(e.id, f.id, 9);
graph.addConnection(e.id, d.id, 6);
graph.addConnection(f.id, e.id, 9);
graph.addConnection(f.id, c.id, 2);
graph.addConnection(f.id, a.id, 14);

function dijkstra(graph, source, target) { // eslint-disable-line
  // create queue
  const queue = graph.vertices.slice().map(i => i.id);
  // initalizing
  graph.reset();
  graph.getVertex(source).dist = 0; // eslint-disable-line
  // algorithm
  while (queue.length) {
    // find a vertex that has min dist
    let u = graph.getVertex(graph.vertices
      .filter(i => queue.includes(i.id))
      .sort((first, second) => first.dist - second.dist)[0].id);
    // find target
    if (u.id === target) {
      const result = [];
      if (u.prev || u.id === source) {
        while (u) {
          result.unshift(u);
          u = u.prev;
        }
      }
      return result;
    }
    // remove u from queue
    {
      const index = queue.indexOf(u.id);
      queue.splice(index, 1);
    }

    for (const v of u.neighbors) {
      if (queue.includes(v.id)) {
        const alt = u.dist + v.weight;
        const orgV = graph.getVertex(v.id);
        if (alt < orgV.dist) {
          orgV.dist = alt;
          orgV.prev = u;
        }
      }
    }
  }
}

dijkstra(graph, 0, 4);

// console.log(dijkstra(graph, 0, 4));
